R/time_bins.R

```diff
@@ -204,59 +191,59 @@ time_bins <-
-      # number of bins, based on bin size
-      nbin <- max(c(round(dur / size), 1))
-      # sort mid ages into bins
-      binlist <-
-        cut(
-          df$mid_ma,
-          breaks = seq(min(df$min_ma), max(df$max_ma), length.out = nbin + 1),
-          include.lowest = T
-        )
-      levels(binlist) <- 1:nbin # levels from 1:nbin for convenience
-      # convert to interval names
-      binlist <-
-        sapply(1:nbin, function(x)
-          df$interval_name[which(binlist == x)])
-      # make list if its only one bin
-      if (!("list" %in% class(binlist)))
-        binlist <- list(binlist)
-      # remove empty bins
-      binlist <-
-        binlist[which(sapply(binlist, function(x)
-          ! (identical(x, character(
-            0
-          )))))]
+
+    if(is.numeric(size) == TRUE){
+      # Update bin size for age range
+      # How many bins should be generated?
+      n_bins <- round((max(df$max_ma)-min(df$min_ma))/size)
+      size <- (max(df$max_ma)-min(df$min_ma))/n_bins
+      #track cumulative sum
+      tracker <- list()
+      for(i in 1:nrow(df)){
+        tracker[[i]] <- rep(NA, length.out = nrow(df))
+        tracker[[i]][i:nrow(df)] <- abs(cumsum(df$duration_myr[i:nrow(df)]) - size)
+      }
+
+      #calculate upper and lower limits for each bin
+      lower <- NULL
+      upper <- NULL
+      count <- 1
+      while(count <= nrow(df)){
+        upper <- append(upper, which.min(tracker[[count]]))
+        lower <- append(lower, (count))
+        count <- which.min(tracker[[count]]) + 1
+      }
 
       #generate bin information
-      bin <- length(binlist):1
-      max_ma <- sapply(binlist, function(x)
-        max(df$max_ma[which(df$interval_name %in% x)]))
-      min_ma <- sapply(binlist, function(x)
-        min(df$min_ma[which(df$interval_name %in% x)]))
-      mid_ma <- (max_ma + min_ma) / 2
+      bin <- length(upper):1
+      max_ma <- df[upper,c("max_ma")]
+      min_ma <- df[lower,c("min_ma")]
+      mid_ma <- (max_ma + min_ma)/2
       duration_myr <- max_ma - min_ma
-      intervals <- sapply(binlist, function(x)
-        toString(x, sep = ", "))
+      intervals <- vector("character")
+      #get interval names
+      for(i in 1:length(upper)){
+        intervals[i] <- toString(df[lower[i]:upper[i], c("interval_name")])
+      }
 
       #generate dataframe
       grouping_rank <- rank
-      df <-
-        cbind.data.frame(bin,
-                         max_ma,
-                         mid_ma,
-                         min_ma,
-                         duration_myr,
-                         grouping_rank,
-                         intervals)
-      rm(bin,
+      df <- cbind.data.frame(bin,
+                             max_ma,
+                             mid_ma,
+                             min_ma,
+                             duration_myr,
+                             grouping_rank,
+                             intervals)
+
+      #remove objects
+      rm(n_bins,
```

Comment:
I don't think we need to manually handle variable cleanup. I believe R has smart variable cleanup, right? It's not hurting anyone as-is, but it might be slightly slower to do it manually, I'm not entirely sure.